### PR TITLE
Allow CSE of immutable loads across stores #9562 (port upstream PR#9562)

### PR DESCRIPTION
**Pull request by gretay-js**
see https://github.com/ocaml/ocaml/pull/9562

### CODE REVIEW

File: backend/mach.ml [hunk 1]
```diff
@@ -35,6 +35,8 @@ type integer_operation =
 
 type float_comparison = Cmm.float_comparison
 
+type mutable_flag = Immutable | Mutable
```
Comment (xclerc) [inline backend/mach.ml:38]: (I am not entirely sure why a new type is introduced;
the other file uses `Asttypes. mutable_flag`, with the
very same definition.)

File: backend/selectgen.ml [hunk 2]
```diff
@@ -156,6 +156,10 @@ let env_close_regions env rs =
   in
   aux None env.regions rs
 
+let select_mutable_flag : Asttypes.mutable_flag -> Mach.mutable_flag = function
```
Comment (xclerc) [inline backend/selectgen.ml:159]: (If we keep the `Mach.mutable_flag` type explicitly
as a non-synonym, maybe move this conversion
function to the `Mach` module - I am afraid no one
would look here for such a function.)

**Review by xclerc — APPROVED** [review]

File: backend/mach.ml (hunk 1 above)
Comment (mshinwell) [reply backend/mach.ml:38]: I think it's generally better not to pilfer one IR's types to use in another IR: it makes it easier to change in the future if one of the IRs needs to diverge.

File: backend/cfg/cfg_equivalence.ml [hunk 3]
```diff
@@ -208,8 +208,10 @@ let check_operation : location -> Cfg.operation -> Cfg.operation -> unit =
     ()
   | Stackoffset expected, Stackoffset result when Int.equal expected result ->
     ()
-  | Load (expected_mem, expected_arch_mode), Load (result_mem, result_arch_mode)
+  | ( Load (expected_mem, expected_arch_mode, expected_mut),
+      Load (result_mem, result_arch_mode, result_mut) )
     when Cmm.equal_memory_chunk expected_mem result_mem
+         && expected_mut = result_mut
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply backend/cfg/cfg_equivalence.ml:214]: Yeah, but this probably throw-away code, so let's allow it.

File: backend/selectgen.ml (hunk 2 above)
Comment (mshinwell) [reply backend/selectgen.ml:159]: Maybe later.  I'd like to get this one merged.

**Review by mshinwell — APPROVED** [review]

**Pull request merged**